apis/ctrlmesh/proto/ctrlmesh_internal.go

```diff
@@ -17,117 +17,146 @@ limitations under the License.
+			SubsetPublicResources: r.SubsetPublicResources,
 		}
-	}
-	if len(route.Subsets) > 0 {
-		if err := json.Unmarshal([]byte(route.Subsets), &ir.Subsets); err != nil {
-			return err
+		for _, subsetLimit := range r.SubsetLimits {
+			is.RouteInternal.SubsetLimits = append(is.RouteInternal.SubsetLimits, &InternalSubsetLimit{
+				Subset: subsetLimit.Subset,
+				Limits: convertProtoMatchLimitRuleToInternal(subsetLimit.Limits),
+			})
 		}
 	}
-	ir.SensSubsetNamespaces = nil
-	for _, s := range route.SensSubsetNamespaces {
-		ir.SensSubsetNamespaces = append(ir.SensSubsetNamespaces, InternalSensSubsetNamespaces{Name: s.Name, Namespaces: sets.NewString(s.Namespaces...)})
-	}
-	if len(route.GlobalExcludeNamespaces) > 0 {
-		ir.GlobalExcludeNamespaces = sets.NewString(route.GlobalExcludeNamespaces...)
-	}
-	return nil
+
+	return is
 }
 
-func (ir *InternalRoute) Encode() *Route {
-	if ir == nil {
-		return nil
-	}
-	if ir.GlobalLimits == nil {
-		ir.GlobalLimits = make([]ctrlmeshv1alpha1.MatchLimitSelector, 0)
-	}
-	if ir.SubRules == nil {
-		ir.SubRules = make([]ctrlmeshv1alpha1.VirtualAppRouteSubRule, 0)
-	}
-	if ir.Subsets == nil {
-		ir.Subsets = make([]ctrlmeshv1alpha1.VirtualAppSubset, 0)
-	}
-	var sensSubsetNamespaces []*SensSubsetNamespaces
-	for _, s := range ir.SensSubsetNamespaces {
-		sensSubsetNamespaces = append(sensSubsetNamespaces, &SensSubsetNamespaces{Name: s.Name, Namespaces: s.Namespaces.List()})
-	}
-	return &Route{
-		Subset:                  ir.Subset,
-		GlobalLimits:            dumpJSON(ir.GlobalLimits),
-		SubRules:                dumpJSON(ir.SubRules),
-		Subsets:                 dumpJSON(ir.Subsets),
-		SensSubsetNamespaces:    sensSubsetNamespaces,
-		GlobalExcludeNamespaces: ir.GlobalExcludeNamespaces.List(),
+func convertProtoMatchLimitRuleToInternal(limits []*MatchLimitRuleV1) []*InternalMatchLimitRule {
+	var internalLimits []*InternalMatchLimitRule
+	for _, limit := range limits {
+		internalLimits = append(internalLimits, &InternalMatchLimitRule{Namespaces: sets.NewString(limit.Namespaces...), Resources: limit.Resources})
 	}
+	return internalLimits
 }
 
-func (ir *InternalRoute) IsDefaultEmpty() bool {
-	return ir.Subset == "" && ir.GlobalExcludeNamespaces.Len() == 0 && len(ir.SensSubsetNamespaces) == 0
+type InternalSpec struct {
+	*ProxySpecV1
+	RouteInternal *InternalRoute
 }
 
-func (ir *InternalRoute) IsNamespaceMatch(ns string) bool {
-	subset, ok := ir.DetermineNamespaceSubset(ns)
+type InternalRoute struct {
+	Subset                string
+	GlobalLimits          []*InternalMatchLimitRule
+	SubsetLimits          []*InternalSubsetLimit
+	SubsetPublicResources []*APIGroupResourceV1
+}
+
+type InternalSubsetLimit struct {
+	Subset string
+	Limits []*InternalMatchLimitRule
+}
+
+type InternalMatchLimitRule struct {
+	Namespaces sets.String
+	// TODO: parse from objectSelector string in protobuf
+	// ObjectSelector    *metav1.LabelSelector
+	Resources []*APIGroupResourceV1
```

Comment:
For a special `Subset`. When we want to customize a rule to limit some resources(e.g. `v1.Pod`), and limit other resources with a diff rule. Should i specify all other resources in `APIGroupResource`?

Reply:
In a rule you can write multiple `match` and each of them can have its specific resources.

For example:

```yaml
  route:
    subRules:
    - name: canary-rule
      match:
      - namespaceSelector: # or objectSelector
          # ...
        resources:
          # v1.Pod
      - namespaceSelector: # or objectSelector
          # ...
        resources:
          # other resources
```